ENH: _lib: add common machinery for low-level callback functions

pv opened the pull request:
Replace the various approaches to combined Python and low-level compiled callback functions (e.g. in integrate.quad, ndimage, ...) with something better. 

We need this part of a FFI layer in Scipy, because the FFI solutions in Python are not standard, and we want to interoperate with all of them. Moreover, FFI is not standard even across Scipy, and we want a reusable solution.
Supported are:
- pure-python callbacks
- cython "api" exported C functions
- ctypes function pointers
- pycapsule'd function pointers
- cffi function pointers.

Caller/thunk code can be implemented in C or Cython.

I ended up with a design where the callback functions need to be explicitly wrapped by the user in `LowLevelCallable` object. This is because parsing e.g. ctypes callback signatures can take a few µs, so users may want to "precompile" them. Another reason is that if sometime in the future, PyPy becomes more important and Scipy switches to cffi etc, changing the design probably is easier.

The specific use cases in integrate.quad and ndimage still handle also the current "legacy" formats, to preserve backward compat.

Fixes gh-4831, gh-5002
- [x] test reentrancy/threadsafety rigorously
- [x] write cython .pxd wrapper for ccallback.h
- [x] decide if the (module, "name") cython syntax makes sense
- [x] decide if the (func, user_data) syntax makes sense
- [x] convert ndimage callbacks
- [x] maybe move everything from `_ccallback.c` to Cython? --- no need to do that in the end, leave some test code in C as a coding example

pv commented:
I think this is more or less ready, comments welcome.
More controversial point may be where to put `LowLevelCallable` --- I put it in the top level namespace.

nmayorov commented:
It looks like a really great work.

I wonder if it would be possible to make an extended tutorial for `LowLevelCallable` where usage of each option from `ctypes, cffi, Cython, or contained in Python `PyCapsule` objects.` will be shown on a single problem? It will be helpful for people who not well familiar with Python-C interaction (like myself).

rgommers commented:
> More controversial point may be where to put LowLevelCallable --- I put it in the top level namespace.

We never put anything there, but in this case I think I agree - there's no other obvious place.

rgommers commented:
> I wonder if it would be possible to make an extended tutorial

@nmayorov the last commit adds documentation which looks quite nice to me. See https://926-1460385-gh.circle-artifacts.com/0//home/ubuntu/scipy/doc/build/html-scipyorg/ccallback.html

rgommers commented:
Looks impressive @pv. Reviewing this PR is nontrivial - it's massive and pretty low-level. I'm tempted to just start using it and see if anything unexpected happens. Starting at the top of the diff somehow doesn't make too much sense to me. And I'll ping the people on gh-4831 who were interested.

pv commented:
For reviewing, I would suggest starting by looking at:
.
1. The new C code in __quadpack.h, compared to the old.
2. The changes in C code in nd_image.c
3. The contents of ccallback.h
4. The suggested usage in the examples e.g. tutorial
5. Think about whether the LowLevelCallable should be put inside quad() 
so that you could just pass in cffi/ctypes callables. But I think the 
current way is a good approach, because it solves the problem of how to 
pass in the void* pointer to additional data without having to add new 
keyword arguments to functions.
.
The rest of the code is mostly for testing the feature.
.
As I see it, the alternative to this would be converting our wrapper 
codes to cffi --- it appears to me to be the best of the FFI options for 
this purpose. However, I don't see how it could work together with Cython.

person142 commented:
Would it make sense to collect the examples into a single new tutorial page? Once this is in I imagine use of it is going to proliferate (seems like it's already in the works for ODE solvers), and it might save replication to have an authoritative guide to point to.

pv commented:
No, I think "cdef api" does the same magic.

pv commented:
Maybe, but there are limited examples on which to demonstrate it currently.

person142 commented:
> No, I think "cdef api" does the same magic.

So it does. That's a good trick to know about.

dhirschfeld commented:
I tried building the branch but with no luck :( First I had to downgrade setuptools because of incorrect path quoting (IIUC) after which it got a lot further but finally failed  with the below errors:
```
building 'scipy.ndimage._nd_image' extension
compiling C sources
creating build\temp.win-amd64-3.5\Release\scipy\ndimage
creating build\temp.win-amd64-3.5\Release\scipy\ndimage\src
C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\x86_amd64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -Iscipy\ndimage\src -IC:\bin\Anaconda3\lib\site-packages\numpy\core\include -Iscipy\_lib\src -IC:\bin\Anaconda3\lib\site-packages\numpy\core\include -IC:\bin\Anaconda3\include -IC:\bin\Anaconda3\include -I"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\INCLUDE" -I"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\ATLMFC\INCLUDE" -I"C:\Program Files (x86)\Windows Kits\10\include\10.0.10240.0\ucrt" -I"C:\Program Files (x86)\Windows Kits\NETFXSDK\4.6.1\include\um" -I"C:\Program Files (x86)\Windows Kits\8.1\include\\shared" -I"C:\Program Files (x86)\Windows Kits\8.1\include\\um" -I"C:\Program Files (x86)\Windows Kits\8.1\include\\winrt" -I"C:\Program Files (x86)\IntelSWTools\compilers_and_libraries\windows\compiler\include" -I"C:\Program Files (x86)\IntelSWTools\compilers_and_libraries\windows\compiler\include\intel64" -I"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\INCLUDE" -I"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\ATLMFC\INCLUDE" -I"C:\Program Files (x86)\Windows Kits\10\include\10.0.10240.0\ucrt" -I"C:\Program Files (x86)\Windows Kits\NETFXSDK\4.6.1\include\um" -I"C:\Program Files (x86)\Windows Kits\8.1\include\\shared" -I"C:\Program Files (x86)\Windows Kits\8.1\include\\um" -I"C:\Program Files (x86)\Windows Kits\8.1\include\\winrt" -I"C:\Program Files (x86)\IntelSWTools\compilers_and_libraries\windows\mpi\intel64\bin\..\..\intel64\include" -I"C:\Program Files (x86)\IntelSWTools\compilers_and_libraries\windows\mkl\include" -I"C:\Program Files (x86)\IntelSWTools\compilers_and_libraries\windows\tbb\bin\..\include" -I"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\INCLUDE" -I"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\ATLMFC\INCLUDE" -I"C:\Program Files (x86)\Windows Kits\10\include\10.0.10240.0\ucrt" -I"C:\Program Files (x86)\Windows Kits\NETFXSDK\4.6.1\include\um" -I"C:\Program Files (x86)\Windows Kits\8.1\include\\shared" -I"C:\Program Files (x86)\Windows Kits\8.1\include\\um" -I"C:\Program Files (x86)\Windows Kits\8.1\include\\winrt" -I"C:\Program Files (x86)\IntelSWTools\compilers_and_libraries\windows\compiler\include" -I"C:\Program Files (x86)\IntelSWTools\compilers_and_libraries\windows\compiler\include\intel64" -I"C:\Program Files (x86)\IntelSWTools\compilers_and_libraries\windows\mpi\intel64\bin\..\..\intel64\include" -I"C:\Program Files (x86)\IntelSWTools\compilers_and_libraries\windows\mkl\include" -I"C:\Program Files (x86)\IntelSWTools\compilers_and_libraries\windows\tbb\bin\..\include" /Tcscipy\ndimage\src\nd_image.c /Fobuild\temp.win-amd64-3.5\Release\scipy\ndimage\src\nd_image.obj
nd_image.c
c:\bin\anaconda3\lib\site-packages\numpy\core\include\numpy\npy_1_7_deprecated_api.h(12) : Warning Msg: Using deprecated NumPy API, disable it by #defining NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
C:\bin\Anaconda3\lib\site-packages\numpy\core\include\numpy/npy_3kcompat.h(179): warning C4244: '=': conversion from 'Py_ssize_t' to 'int', possible loss of data
scipy\_lib\src\ccallback.h(133): error C2061: syntax error: identifier 'jmp_buf'
scipy\_lib\src\ccallback.h(140): error C2059: syntax error: '}'
scipy\_lib\src\ccallback.h(260): warning C4133: 'function': incompatible types - from 'PyTypeObject *' to 'PyObject *'
scipy\_lib\src\ccallback.h(275): error C2037: left of 'py_function' specifies undefined struct/union 'ccallback'
scipy\_lib\src\ccallback.h(276): error C2037: left of 'c_function' specifies undefined struct/union 'ccallback'
scipy\_lib\src\ccallback.h(277): error C2037: left of 'user_data' specifies undefined struct/union 'ccallback'
scipy\_lib\src\ccallback.h(278): error C2037: left of 'signature_index' specifies undefined struct/union 'ccallback'
scipy\_lib\src\ccallback.h(284): error C2037: left of 'py_function' specifies undefined struct/union 'ccallback'
scipy\_lib\src\ccallback.h(285): error C2037: left of 'c_function' specifies undefined struct/union 'ccallback'
scipy\_lib\src\ccallback.h(286): error C2037: left of 'user_data' specifies undefined struct/union 'ccallback'
scipy\_lib\src\ccallback.h(287): error C2037: left of 'signature_index' specifies undefined struct/union 'ccallback'
scipy\_lib\src\ccallback.h(307): error C2037: left of 'signature_index' specifies undefined struct/union 'ccallback'
scipy\_lib\src\ccallback.h(308): error C2037: left of 'signature_index' specifies undefined struct/union 'ccallback'
scipy\_lib\src\ccallback.h(325): error C2037: left of 'py_function' specifies undefined struct/union 'ccallback'
scipy\_lib\src\ccallback.h(326): error C2037: left of 'c_function' specifies undefined struct/union 'ccallback'
scipy\_lib\src\ccallback.h(327): error C2037: left of 'user_data' specifies undefined struct/union 'ccallback'
scipy\_lib\src\ccallback.h(335): error C2037: left of 'prev_callback' specifies undefined struct/union 'ccallback'
scipy\_lib\src\ccallback.h(339): error C2037: left of 'prev_callback' specifies undefined struct/union 'ccallback'
scipy\_lib\src\ccallback.h(354): error C2037: left of 'prev_callback' specifies undefined struct/union 'ccallback'
scipy\_lib\src\ccallback.h(355): error C2037: left of 'prev_callback' specifies undefined struct/union 'ccallback'
scipy\_lib\src\ccallback.h(355): error C2198: 'ccallback__set_thread_local': too few arguments for call
scipy\_lib\src\ccallback.h(357): error C2037: left of 'prev_callback' specifies undefined struct/union 'ccallback'
scipy\_lib\src\ccallback.h(358): error C2037: left of 'c_function' specifies undefined struct/union 'ccallback'
scipy\_lib\src\ccallback.h(359): error C2037: left of 'py_function' specifies undefined struct/union 'ccallback'
scipy\ndimage\src\nd_image.c(284): error C2037: left of 'info_p' specifies undefined struct/union 'ccallback'
scipy\ndimage\src\nd_image.c(296): error C2037: left of 'py_function' specifies undefined struct/union 'ccallback'
scipy\ndimage\src\nd_image.c(296): error C2198: 'PyObject_Call': too few arguments for call
scipy\ndimage\src\nd_image.c(320): error C2079: 'callback' uses undefined struct 'ccallback'
scipy\ndimage\src\nd_image.c(327): error C2224: left of '.py_function' must have struct/union type
scipy\ndimage\src\nd_image.c(328): error C2224: left of '.c_function' must have struct/union type
scipy\ndimage\src\nd_image.c(360): warning C4133: 'function': incompatible types - from 'int *' to 'ccallback_t *'
scipy\ndimage\src\nd_image.c(365): error C2224: left of '.py_function' must have struct/union type
scipy\ndimage\src\nd_image.c(368): error C2224: left of '.info_p' must have struct/union type
scipy\ndimage\src\nd_image.c(373): error C2224: left of '.c_function' must have struct/union type
scipy\ndimage\src\nd_image.c(374): error C2224: left of '.user_data' must have struct/union type
scipy\ndimage\src\nd_image.c(382): error C2224: left of '.py_function' must have struct/union type
scipy\ndimage\src\nd_image.c(382): error C2224: left of '.c_function' must have struct/union type
scipy\ndimage\src\nd_image.c(383): warning C4133: 'function': incompatible types - from 'int *' to 'ccallback_t *'
scipy\ndimage\src\nd_image.c(396): error C2037: left of 'info_p' specifies undefined struct/union 'ccallback'
scipy\ndimage\src\nd_image.c(407): error C2037: left of 'py_function' specifies undefined struct/union 'ccallback'
scipy\ndimage\src\nd_image.c(407): error C2198: 'PyObject_Call': too few arguments for call
scipy\ndimage\src\nd_image.c(428): error C2079: 'callback' uses undefined struct 'ccallback'
scipy\ndimage\src\nd_image.c(435): error C2224: left of '.py_function' must have struct/union type
scipy\ndimage\src\nd_image.c(436): error C2224: left of '.c_function' must have struct/union type
scipy\ndimage\src\nd_image.c(468): warning C4133: 'function': incompatible types - from 'int *' to 'ccallback_t *'
scipy\ndimage\src\nd_image.c(473): error C2224: left of '.py_function' must have struct/union type
scipy\ndimage\src\nd_image.c(476): error C2224: left of '.info_p' must have struct/union type
scipy\ndimage\src\nd_image.c(481): error C2224: left of '.c_function' must have struct/union type
scipy\ndimage\src\nd_image.c(482): error C2224: left of '.user_data' must have struct/union type
scipy\ndimage\src\nd_image.c(489): error C2224: left of '.py_function' must have struct/union type
scipy\ndimage\src\nd_image.c(489): error C2224: left of '.c_function' must have struct/union type
scipy\ndimage\src\nd_image.c(490): warning C4133: 'function': incompatible types - from 'int *' to 'ccallback_t *'
scipy\ndimage\src\nd_image.c(572): error C2037: left of 'info_p' specifies undefined struct/union 'ccallback'
scipy\ndimage\src\nd_image.c(588): error C2037: left of 'py_function' specifies undefined struct/union 'ccallback'
scipy\ndimage\src\nd_image.c(588): error C2198: 'PyObject_Call': too few arguments for call
scipy\ndimage\src\nd_image.c(614): error C2079: 'callback' uses undefined struct 'ccallback'
scipy\ndimage\src\nd_image.c(621): error C2224: left of '.py_function' must have struct/union type
scipy\ndimage\src\nd_image.c(622): error C2224: left of '.c_function' must have struct/union type
scipy\ndimage\src\nd_image.c(658): warning C4133: 'function': incompatible types - from 'int *' to 'ccallback_t *'
scipy\ndimage\src\nd_image.c(663): error C2224: left of '.py_function' must have struct/union type
scipy\ndimage\src\nd_image.c(666): error C2224: left of '.info_p' must have struct/union type
scipy\ndimage\src\nd_image.c(671): error C2224: left of '.c_function' must have struct/union type
scipy\ndimage\src\nd_image.c(672): error C2224: left of '.user_data' must have struct/union type
scipy\ndimage\src\nd_image.c(682): error C2224: left of '.py_function' must have struct/union type
scipy\ndimage\src\nd_image.c(682): error C2224: left of '.c_function' must have struct/union type
scipy\ndimage\src\nd_image.c(683): warning C4133: 'function': incompatible types - from 'int *' to 'ccallback_t *'
error: Command "C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\x86_amd64\cl.exe /c /nologo /Ox /W3 /GL /DNDEBUG /MD -Iscipy\ndimage\src -IC:\bin\Anaconda3\lib\site-packages\numpy\core\include -Iscipy\_lib\src -IC:\bin\Anaconda3\lib\site-packages\numpy\core\include -IC:\bin\Anaconda3\include -IC:\bin\Anaconda3\include -I"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\INCLUDE" -I"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\ATLMFC\INCLUDE" -I"C:\Program Files (x86)\Windows Kits\10\include\10.0.10240.0\ucrt" -I"C:\Program Files (x86)\Windows Kits\NETFXSDK\4.6.1\include\um" -I"C:\Program Files (x86)\Windows Kits\8.1\include\\shared" -I"C:\Program Files (x86)\Windows Kits\8.1\include\\um" -I"C:\Program Files (x86)\Windows Kits\8.1\include\\winrt" -I"C:\Program Files (x86)\IntelSWTools\compilers_and_libraries\windows\compiler\include" -I"C:\Program Files (x86)\IntelSWTools\compilers_and_libraries\windows\compiler\include\intel64" -I"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\INCLUDE" -I"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\ATLMFC\INCLUDE" -I"C:\Program Files (x86)\Windows Kits\10\include\10.0.10240.0\ucrt" -I"C:\Program Files (x86)\Windows Kits\NETFXSDK\4.6.1\include\um" -I"C:\Program Files (x86)\Windows Kits\8.1\include\\shared" -I"C:\Program Files (x86)\Windows Kits\8.1\include\\um" -I"C:\Program Files (x86)\Windows Kits\8.1\include\\winrt" -I"C:\Program Files (x86)\IntelSWTools\compilers_and_libraries\windows\mpi\intel64\bin\..\..\intel64\include" -I"C:\Program Files (x86)\IntelSWTools\compilers_and_libraries\windows\mkl\include" -I"C:\Program Files (x86)\IntelSWTools\compilers_and_libraries\windows\tbb\bin\..\include" -I"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\INCLUDE" -I"C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\ATLMFC\INCLUDE" -I"C:\Program Files (x86)\Windows Kits\10\include\10.0.10240.0\ucrt" -I"C:\Program Files (x86)\Windows Kits\NETFXSDK\4.6.1\include\um" -I"C:\Program Files (x86)\Windows Kits\8.1\include\\shared" -I"C:\Program Files (x86)\Windows Kits\8.1\include\\um" -I"C:\Program Files (x86)\Windows Kits\8.1\include\\winrt" -I"C:\Program Files (x86)\IntelSWTools\compilers_and_libraries\windows\compiler\include" -I"C:\Program Files (x86)\IntelSWTools\compilers_and_libraries\windows\compiler\include\intel64" -I"C:\Program Files (x86)\IntelSWTools\compilers_and_libraries\windows\mpi\intel64\bin\..\..\intel64\include" -I"C:\Program Files (x86)\IntelSWTools\compilers_and_libraries\windows\mkl\include" -I"C:\Program Files (x86)\IntelSWTools\compilers_and_libraries\windows\tbb\bin\..\include" /Tcscipy\ndimage\src\nd_image.c /Fobuild\temp.win-amd64-3.5\Release\scipy\ndimage\src\nd_image.obj" failed with exit status 2
```

pv commented:
@dhirschfeld: may be just missing `#include <setjmp.h>`, added in 56f0925

dhirschfeld commented:
Thanks @pv - that fixed the compilation so I can now test this branch on py35/win64.

I'm very interested in this branch as I have an expensive integral which I'm hoping to speedup using the numba [`cfunc`](http://numba.pydata.org/numba-doc/dev/user/cfunc.html) functionality. I haven't been able to make full use of it so far because `quad` hasn't allowed passing in extra parameters correctly when using the `ctypes` interface (IIUC).

```python
In [4]: from scipy.integrate import quad

In [5]: def f1(x, mu):
   ...:     return np.exp(-x**2)/(1+np.exp(-(x-mu)))

In [6]: quad(f1, -np.inf, np.inf, args=(1.1,))
Out[6]: (0.47866266846942096, 5.656862062202448e-09)

In [7]: quad(f1, -np.inf, np.inf, args=(1.9,))
Out[7]: (0.26575640518956256, 2.7307507646266408e-09)
```
I implemented this using numbas `cfunc` as follows:
```python
In [58]: from numba import types
    ...: from numba import carray, cfunc, float64
    ...: from scipy import LowLevelCallable

In [59]: @cfunc(float64(float64, types.voidptr))
    ...: def f2(x, ptr_mu):
    ...:     mu = carray(ptr_mu, (1,), dtype=float64)[0]
    ...:     return np.exp(-x**2)/(1+np.exp(-(x-mu)))

In [60]: mu = ctypes.c_double(1.1)
    ...: mu_ptr = ctypes.cast(ctypes.pointer(mu), ctypes.c_void_p)
    ...: f = LowLevelCallable(f2.ctypes, mu_ptr)

In [61]: quad(f, -np.inf, np.inf)
Out[61]: (0.47866266846942096, 5.656862062202448e-09)

In [62]: mu = ctypes.c_double(1.9)
    ...: mu_ptr = ctypes.cast(ctypes.pointer(mu), ctypes.c_void_p)
    ...: f = LowLevelCallable(f2.ctypes, mu_ptr)

In [63]: quad(f, -np.inf, np.inf)
Out[63]: (0.26575640518956256, 2.7307507646266408e-09)
```
...with a resulting 16x speedup!
```python
In [73]: %timeit quad(f, -np.inf, np.inf)
10000 loops, best of 3: 53.4 µs per loop

In [74]: %timeit quad(f1, -np.inf, np.inf, args=(1.9,))
1000 loops, best of 3: 859 µs per loop

In [75]: quad(f1, -np.inf, np.inf, args=(1.9,))[0] == quad(f, -np.inf, np.inf)[0]
Out[75]: True
```
🎉

dhirschfeld commented:
...since it's slightly non-trivial it might make for a good example - but that's possibly beter located in the numba docs?

**cc:** @pitrou

dhirschfeld commented:
When there's more than one extra argument to the callback it gets even more convoluted:
```python
In [46]: def f1(x, mu, alpha):
    ...:     return np.exp(-alpha*x**2)/(1+np.exp(-(x-mu)))

In [47]: mu = 1.9
    ...: alpha = 0.77
    ...: args = np.asarray((mu, alpha), dtype=[('mu', np.float64),('alpha', np.float64)])

In [48]: quad(f1, -np.inf, np.inf, args=(mu, alpha))
Out[48]: (0.3138771115843293, 3.516492370046489e-09)
In [49]: dtype = args.dtype

In [50]: @cfunc(float64(float64, types.voidptr))
    ...: def f2(x, args):
    ...:     args = carray(args, (1,), dtype=dtype)
    ...:     mu = args.mu[0]
    ...:     alpha = args.alpha[0]
    ...:     return np.exp(-alpha*x**2)/(1+np.exp(-(x-mu)))

In [51]: f = LowLevelCallable(f2.ctypes, ctypes.c_voidp(args.ctypes.data))
    ...: quad(f, -np.inf, np.inf)
Out[51]: (0.31387711158432935, 3.5164923763348615e-09)
In [52]: %timeit quad(f1, -np.inf, np.inf, args=(mu, alpha))
1000 loops, best of 3: 661 µs per loop

In [53]: %timeit quad(f, -np.inf, np.inf)
10000 loops, best of 3: 41.8 µs per loop
```

AFAICS, since there's no concept of a void pointer to a struct of data in python, it's fairly awkward to use with e.g. numba.

I'm wondering if, instead of the signature `LowLevelCallable(function, void*)` it was instead `LowLevelCallable(function, *args)` - if your c-func did actually take a `void*` then you could pass that as the single arg, otherwise you could pass whatever types the callback was expecting. Would that even be possible?

pv commented:
Variadic functions are not doable in C without dynamic code generation.
So the answer is no, it's not possible.

pitrou commented on scipy/_lib/_ccallback.py:
```diff
@@ -0,0 +1,209 @@
+
+try:
+    import cffi
+    ffi = cffi.FFI()
```
Beware that this can be costly:
```
>>> %time import cffi; cffi.FFI().CData
CPU times: user 24 ms, sys: 0 ns, total: 24 ms
Wall time: 21.7 ms
```

pitrou commented on scipy/_lib/src/_test_ccallback.c:
```diff
@@ -0,0 +1,363 @@
+ * This also is an internal "best-practices" code example on how to write
+ * low-level callback code.
+ *
+ * The *thunk_simple* function shows how to write the wrapper callback
```
Who would need to write such a thunk function?

pitrou commented on scipy/_lib/src/_test_ccallback.c:
```diff
@@ -0,0 +1,363 @@
+        PyGILState_Release(state);
+    }
+    else {
+        if (callback->signature_index == 0) {
```
Hmm, this looks obscure to me.

pitrou commented on scipy/_lib/src/_test_ccallback.c:
```diff
@@ -0,0 +1,363 @@
+    }
+
+    if (error) {
+        *error_flag = 1;
```
I'm a bit surprised that `*error_flag` isn't set to 0 explicitly otherwise. Is it best practice?

pv commented:
This is test code. The thunk signature is whatever best fits the code 
that needs the callback. E.g. in QUADPACK the thunk signature is 
determined by what QUADPACK expects.

pitrou commented on scipy/_lib/src/ccallback.h:
```diff
@@ -0,0 +1,363 @@
+        lowlevelcallable_type = (PyTypeObject *)PyObject_GetAttrString(module, "LowLevelCallable");
+        if (lowlevelcallable_type == NULL) {
+            goto error;
+        }
```
You may want to decref the module object after this.

pitrou commented on scipy/_lib/src/ccallback.h:
```diff
@@ -0,0 +1,363 @@
+    }
+
+    if ((flags & CCALLBACK_PARSE) && !PyObject_TypeCheck(callback_obj, lowlevelcallable_type)) {
+        /* Parse callback */
```
Out of curiosity, why is this conditioned on CCALLBACK_PARSE?

pitrou commented on scipy/_lib/src/ccallback.h:
```diff
@@ -0,0 +1,363 @@
+            goto error;
+        }
+
+        callback_obj = callback_obj2;
```
If I'm not mistaken, `callback_obj2` is a new reference, while `callback_obj` was borrowed.

rgommers replied:
@pv this comment still seems open, not clear to me that it was addressed

pitrou commented on scipy/_lib/src/ccallback.h:
```diff
@@ -0,0 +1,363 @@
+        callback->signature_index = -1;
+    }
+    else if (PyObject_TypeCheck(callback_obj, lowlevelcallable_type) &&
+             PyTuple_Check(callback_obj) &&
```
I'm probably misreading this but... how can `callback_obj` be a tuple and a LowLevelCallable instance at the same time? (unless LowLevelCallable is a tuple subclass)

pv commented:
CCALLBACK_PARSE exists only for supporting legacy functionality in 
quad(), passing in raw ctypes objects, which I decided not to enable by 
default.

pv commented:
LowLevelCallable is a tuple subclass to make it immutable.

pitrou commented on scipy/_lib/src/ccallback.h:
```diff
@@ -0,0 +1,363 @@
+    }
+    else if (capsule != NULL ||
+             PyObject_TypeCheck(callback_obj, lowlevelcallable_type) &&
+             PyTuple_Check(callback_obj) &&
```
Same comment as above.

pitrou commented on scipy/integrate/tests/test_quadpack.py:
```diff
@@ -31,21 +25,24 @@ def assert_quad(value_and_err, tabled_value, errTol=1.5e-8):
         if sys.platform == 'win32':
             if sys.version_info < (3, 5):
-                file = ctypes.util.find_msvcrt()
+                files = [ctypes.util.find_msvcrt()]
```
For the record, to avoid such murking around trying to find `libm`, you could compile your own wrapper function simply calling `sin()` (that's what we do in Numba).

pitrou commented:
Hi,
I'm a bit worried by the decision to compare C signatures as strings. In Numba, for example, there is no separate notion of a `npy_intp` type. A `numba.types.intp` is simply a `int32` or `int64` depending on the machine's width (`np.intp` works the same way). Therefore I'm not sure if passing a Numba cfunc will always be possible (or, at least, easy) depending on which signatures a given ccallback-consuming function accepts.

pv commented:
I think the problem you see is whether this should be specified on the 
API level or ABI level. The string repr is fine for the former, and is 
very cheap for CFFI and Cython.
.
An ABI level spec (base type + size) would involve quite a bit more 
involved parsing, and needs to be aware of the type sizes for the C 
compiler used.

pv commented:
The question then mainly is, who does the extra legwork of figuring out 
the API--ABI mapping --- the user, or ccallback?

pitrou commented:
You're right that ABI level signature matching would be much more cumbersome. Is it necessary to accept several different signatures for a single caller, though?

pv commented:
It's possible to insert #ifdef's that add additional signatures 
normalizing intp to C base types, to the accepted signature lists. 
ctypes (via which Numba goes) knows the API-ABI mapping, so unless you 
define custom types, that would probably be enough to make it work in 
this case.

pv commented:
Addressed review comments.

The API vs. ABI decision needs to be made to some direction, and I think the one requiring least work is fine. I think why API level is least work, is that the code here is interfacing with the various c-ffi libraries Python has, and those provide information on the API level. I would guess broad compatibility is ensured by (i) always including possible signatures with base C types only --- which should ensure eg. Numba always works, (ii) allowing the user to override the signature manually.

pitrou commented on doc/source/tutorial/ndimage.rst:
```diff
@@ -1759,134 +1739,218 @@ parameters ``output_coordinates`` and ``input_coordinates`` play the
+
+.. rubric:: Numba
+
+Numba_ provides a way to write low-level functions easily in Python.
```
Wow, thank you for taking the time to add this example.

pitrou commented:
>  I think why API level is least work, is that the code here is interfacing with the various c-ffi libraries Python has, and those provide information on the API level. I would guess broad compatibility is ensured by (i) always including possible signatures with base C types only --- which should ensure eg. ctypes/Numba always works, (ii) allowing the user to override the signature manually.

I think you're right.

I left a couple of code review comments, I don't know if you noticed them.

pv commented:
I think I mostly corrected the issues pointed out (e.g. refcount 
ownership) in the commit series above.

pv commented:
One remaining question for Numba is that if `sizeof(int) == 
sizeof(long)`, which integer type will int32 be cast to in the ctypes 
interface? (Both types should probably be listed in the signatures in 
this case here, to avoid the issue.)

pitrou commented:
Like Numpy, Numba has a "C int" (named `intc`) but doesn't know "C long".

pv commented:
I guess the question is more on how Numba constructs the ctypes 
interface --- e.g. does numba.types.intc map to ctypes.c_int even if 
sizeof(int) == sizeof(long), and whether numba.types.int32 maps to c_int 
or c_long in this case (if it uses ctypes.c_int32, that maps to the 
"largest" type).

pitrou commented:
There is a single bijective mapping:
```
_FROM_CTYPES = {
    ctypes.c_bool: types.boolean,
    
    ctypes.c_int8:  types.int8,
    ctypes.c_int16: types.int16,
    ctypes.c_int32: types.int32,
    ctypes.c_int64: types.int64,

    ctypes.c_uint8: types.uint8,
    ctypes.c_uint16: types.uint16,
    ctypes.c_uint32: types.uint32,
    ctypes.c_uint64: types.uint64,

    ctypes.c_float: types.float32,
    ctypes.c_double: types.float64,

    ctypes.c_void_p: types.voidptr,
    ctypes.py_object: types.ffi_forced_object,
}
```

Numba's `intc` and `intp` are just aliases to the explicitly-sized types, so `intc` would map back to `ctypes.c_int32` in most cases.

rgommers commented on scipy/_lib/src/ccallback.h:
```diff
@@ -0,0 +1,363 @@
+
+    if (PyCallable_Check(callback_obj)) {
+        /* Python callable */
+        callback->py_function = callback_obj;
```
@pv this comment still seems open, not clear to me that it was addressed

pv commented:
@rgommers: all the life cycle issues have been addressed, as far as I see.

3.1.2017 8.02 "Ralf Gommers" <notifications@github.com> kirjoitti:

*@rgommers* commented on this pull request.
------------------------------

In scipy/_lib/src/ccallback.h <https://github.com/scipy/scipy/pull/6509>:

> +        callback_obj2 = PyObject_CallMethod(lowlevelcallable_type, "_parse_callback",
+                                            "O", callback_obj);
+        if (callback_obj2 == NULL) {
+            goto error;
+        }
+
+        callback_obj = callback_obj2;
+
+        if (PyCapsule_CheckExact(callback_obj)) {
+            capsule = callback_obj;
+        }
+    }
+
+    if (PyCallable_Check(callback_obj)) {
+        /* Python callable */
+        callback->py_function = callback_obj;

@pv <https://github.com/pv> this comment still seems open, not clear to me
that it was addressed

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/scipy/scipy/pull/6509>, or mute the thread
<https://github.com/notifications/unsubscribe-auth/AACI5mfL3o2vyhIRBq2YSHKmPNcrBsFrks5rOeRCgaJpZM4JqN_5>
.

rgommers commented:
Okay, thanks for confirming.

rgommers commented:
Let's get this in then - has been thoroughly reviewed, seems to be in good shape, and would be good to have it in master for a few weeks before branching 0.19.x.

rgommers commented:
Or do you still want to do anything about this comment regarding numba ints:https://github.com/scipy/scipy/pull/4933 ?

pv commented:
I added adfitional signatures for different int sizes, so it should work on
different platforms.

4.1.2017 5.24 ap. "Ralf Gommers" <notifications@github.com> kirjoitti:

> Or do you still want to do anything about this comment regarding numba
> ints:#4933 <https://github.com/scipy/scipy/pull/4933> ?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scipy/scipy/pull/6509#issuecomment-270286841>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AACI5honufxpGMyOl38G3vzpB2aLdslWks5rOxDigaJpZM4JqN_5>
> .
>

rgommers commented:
Okay in it goes. Thanks Pauli! And thanks everyone for the reviews!
